homework/homework2/task3/smart_args.py

```diff
@@ -0,0 +1,81 @@
+from copy import deepcopy
+from functools import update_wrapper
+from inspect import signature
+from typing import Callable, Any
+
+
+class Evaluated:
+    def __init__(self, func: Callable):
+        """
+        Substitutes the default value calculated at the time of the call.
+
+        :param func: a function that will calculate the default value before each call.
+        """
+        if isinstance(func, Isolated):
+            raise ValueError("Isolated cannot be used with Evaluated")
+
+        if len(signature(func).parameters) != 0:
+            raise ValueError("Functions with arguments are not supported by Evaluated")
+
+        self.func = func
+
+
+class Isolated:
+    def __init__(self, arg=None):
+        """
+        Fictitious default value. Make a deep copy while receiving the argument.
+
+        :param arg: the object to be isolated.
+        """
+        if isinstance(arg, Evaluated):
+            raise ValueError("Evaluated cannot be used with Isolated")
+
+        self.arg = arg
+
+
+# Wraps _SmartArgs to allow for deferred calling
+def smart_args(func=None, *, positional_arguments_included: bool = False):
```

Comment:
Please, fix the following case:
```
@smart_args(positional_arguments_included=True)
def support_positionals(x=Isolated()):
    x[0] += 1
    return x

X = [0]
print(support_positionals(X))
```
It should work since you support positional arguments, but I got `KeyError`

Reply:
Removed support for positional arguments due to incorrect implementation. I'll add their support later.